teps/0061-allow-custom-task-to-be-embedded-in-pipeline.md

```diff
@@ -0,0 +1,239 @@
+---
+status: proposed
+title: Allow custom task to be embedded in pipeline
+creation-date: '2021-03-18'
+last-updated: '2021-03-27'
+authors:
+- '@Tomcli'
+- '@litong01'
+- '@ScrapCodes'
+---
+
+# TEP-0061: Allow custom task to be embedded in pipeline
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+  - [Use Cases (optional)](#use-cases-optional)
+- [Requirements](#requirements)
+- [Proposal](#proposal)
+  - [Notes/Caveats (optional)](#notescaveats-optional)
+  - [Risks and Mitigations](#risks-and-mitigations)
+  - [User Experience (optional)](#user-experience-optional)
+  - [Performance (optional)](#performance-optional)
+- [Design Details](#design-details)
+- [Test Plan](#test-plan)
+- [Design Evaluation](#design-evaluation)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Infrastructure Needed (optional)](#infrastructure-needed-optional)
+- [Upgrade &amp; Migration Strategy (optional)](#upgrade--migration-strategy-optional)
+- [References (optional)](#references-optional)
+<!-- /toc -->
+
+## Summary
+
+Tektoncd/Pipeline currently allows custom task to be referenced in pipeline
+resource specification file using [`taskRef`](https://github.com/tektoncd/community/blob/main/teps/0002-custom-tasks.md).
+That feature allows a custom task to be submitted to kubernetes along
+the submission of the [Tektoncd/pipeline](https://github.com/tektoncd/pipeline),
+however, the submission of a custom task is a separate request to Kubernetes.
+If multiple custom tasks use same name, to both Kubernetes and Tektoncd/Pipeline,
+they will be treated as the same task, this behavior can have unintended
+consequences when Tektoncd/Pipeline gets used as a backend with multiple users.
+This problem becomes even greater when new users follow documents such as
+`Get started` which often use same name for task and pipeline. In this environment
+multiple users will step on each other's toes, and produced unintended results.
+
+## Motivation
+
+It is natural for a user to follow ways such as a [Kubernetes Deployment](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/), ReplicaSet, StatefulSet and also
+Tektoncd/Pipeline taskSpec to have a Pipeline with custom tasks embedded.
+Currently TektonCD/Pipeline supports task specifications to be embedded in
+a pipeline for regular task, but not for custom task. If Tektoncd/Pipeline
+also allows a custom task specification to be embedded in a pipeline specification
+then the behavior will be unified with non custom task and there won't be
+any issues in terms of naming conflict when used by multiple users.
+See issue [tektoncd/pipeline#3682](https://github.com/tektoncd/pipeline/issues/3682)
+
+### Goals
+
+1. Allow custom tasks to be embedded in a pipeline specification,
+2. Custom task specification verification/validation should be handed
+over to custom task controllers, custom task specification must not be
```

Comment:
Could we soften the wording of this one a bit?  I think there are approaches here where the custom task controller could specify the validation (e.g. as openapi) and pipelines controller enforces that validation.

We might not want to make this required, or the default expectation, but I am a tad nervous about "only" treating this field as a generic blob which the pipelines controller has no insight into.  It could make future work around conformance and Custom Tasks (for example) considerably more complicated if we've blanket prohibited the pipelines controller from performing any verification / validation.

We could remove this goal and simply leave it as a non-goal for now?

Reply:
Yes, we can move this to non-goal